soroban-env-host/src/host.rs

```diff
@@ -158,8 +158,31 @@ impl Default for DiagnosticLevel {
     }
 }
 
+/// LegacyEpoch defines a set of _past_ (legacy) behaviour that we have logic in
+/// the host to handle, that can be activated by the embedder to replay past
+/// contracts "the same way they ran before".
+///
+/// Legacy epochs are added when some change in behaviour is detected by the
+/// embedder -- because behaviour does not replay correctly -- at which point
+/// the _current_ behaviour is always considered part of `LegacyEpoch:Current`
+/// and a new _earlier_ epoch is added before it, that is named for the
+/// _previous_ behaviour.
+///
+/// For example, if we had a change to hashing that made it much cheaper, when
```

Comment:
nit: I think this example is a bit misleading because IIUC the metering calibration data is configured via a separate mechanism (config settings) and there shouldn't be an issue about e.g. optimizing the function without changing the calibration (or vice versa). I suppose this should be meaningful for the logic changes (e.g. if a function had a bug or if pre-conditions have changed), but you probably have a better idea of what this covers.

Reply:
This can happen if the input or number of iterations of work has changed, while the unit of work (parameterized and stored in config settings) has not. 
Going along with this example, if the unit of work is hashing 1 byte of data, and is a constant cost. With the faster hashing algorithm, it requires hashing fewer bytes of data for the same object, thus fewer iterations. Then the overall cost will be reduced.

Reply:
Ok, I'll make a different example

Reply:
@dmkozh I've updated the comment here with much more detail as well as an example that's more realistic / not related to metering cost differences. Let me know if this is adequate.

Reply:
This looks good, thanks!